server/src/main/java/io/crate/execution/engine/aggregation/impl/TopKAggregation.java

```diff
@@ -198,6 +216,77 @@ public DataType<?> partialType() {
         return new StateType(boundSignature.argTypes().getFirst());
     }
 
+    @Nullable
+    @Override
+    public DocValueAggregator<?> getDocValueAggregator(LuceneReferenceResolver referenceResolver,
+                                                       List<Reference> aggregationReferences,
+                                                       DocTableInfo table,
+                                                       List<Literal<?>> optionalParams) {
+        if (aggregationReferences.size() != 1) {
+            return null;
+        }
+        Reference reference = aggregationReferences.get(0);
+        if (!reference.hasDocValues()) {
+            return null;
+        }
+
+        int limit;
+        if(optionalParams.size() == 1) {
+            limit = (int) optionalParams.getFirst().value();
+        } else {
+            limit = DEFAULT_LIMIT;
+        }
+        return getDocValueAggregator(reference, limit);
+    }
+
+    @Nullable
+    private DocValueAggregator<?> getDocValueAggregator(Reference ref, int limit) {
+        if (!ref.hasDocValues()) {
+            return null;
+        }
+        return switch (ref.valueType().id()) {
+            case ByteType.ID,
+                 ShortType.ID,
+                 IntegerType.ID,
+                 LongType.ID,
+                 TimestampType.ID_WITH_TZ,
+                 TimestampType.ID_WITHOUT_TZ,
+                 FloatType.ID,
+                 DoubleType.ID,
+                 GeoPointType.ID ->
+                new SortedNumericDocValueAggregator<>(
+                    ref.storageIdent(),
+                    (ramAccounting, _, _) -> {
+                        int maxMapSize = maxMapSize(limit);
+                        ramAccounting.addBytes(calculateRamUsage(maxMapSize));
+                        var state = new TopKState(new ItemsSketch<>(maxMapSize), limit);
+                        ramAccounting.addBytes(RamUsageEstimator.shallowSizeOf(state));
```

Comment:
```suggestion
                        var state = new TopKState(new ItemsSketch<>(maxMapSize), limit);
                        ramAccounting.addBytes(RamUsageEstimator.shallowSizeOf(state) + calculateRamUsage(maxMapSize));
```

Or we could probably declare a constant on `TopKState` for the size via `RamUsageEstimator.shallowSizeOfInstance(...)` and do the accounting once before creating the instance.

(Same actually for the non-optimized generic topk version)

And it looks like the lambda could be shared between this variant and the ip/string type variant below?